### DISCUSSION TITLE
Next. js 14 With Mobx

### DISCUSSION THREAD

**Discussion by mawaisismail** [mobxjs/mobx#3805] [Q&A] [answered]
i tried to create state management with the next.js 14 app router with mobx but some how its not working its changing stats in store but not reflect on UI i also added observer but still not working in there any way to solve it ? and i created context to share the store globally. 

Please share if anyone integrates Mobx in Next. js With app router

**Comment by mawaisismail** [comment, marked as answer]
I fount solution for this this is how i implement mobx with next.js 14

**store**
```typescript
export class UiStore {
  count = 0;
  constructor() {
    makeAutoObservable(this);
  }

  increment = () => {
    this.count++;
  };
}

``` 

**group all store** 

```typescript
export const RootStore = {
  uiStore,
};
``` 


**provider**

```typescript
"use client";
import React, { createContext, ReactNode } from "react";
import { RootStore } from "@/store/store";

export const StoreContext = createContext(RootStore);

export const StoreWrapper = ({ children }: { children: ReactNode }) => {
  return (
    <StoreContext.Provider value={RootStore}>{children}</StoreContext.Provider>
  );
};

``` 

**main layout**

```typescript
  <StoreWrapper>{children}</StoreWrapper>


//hook to get stores

import { useContext } from "react";
import { StoreContext } from "@/store/provider";

export const useStores = () => {
  return useContext(StoreContext);
};
``` 


**component where you want to use store**

```typescript
"use client";
import { observer } from "mobx-react";
import { useStores } from "@/hooks/useStore";

const HomePage = observer(() => {
  const {
    uiStore: { count, increment },
  } = useStores();

  return (
    <>
      <div>{count}</div>
      <button onClick={increment}>Increment</button>)
    </>
  );
});

export default HomePage;
``` 

 [More detailed answer](https://github.com/mobxjs/mobx/discussions/3805#discussioncomment-7914969)

**Reply by jymartineau** [reply]
This is awesome, worked right away for me!

**Comment by ideaviewes** [comment]
How to initialize uiStore ？I can not find you initialize it anywhere
`
export const RootStore = {
  uiStore,
};
`
Thank you for your help.

**Reply by mawaisismail** [reply]
Here's complete code

**useStores**

```typescript
import { useContext } from "react";
import { StoreContext } from "@/store/provider";

export const useStores = () => {
  return useContext(StoreContext);
};
``` 

**StoreContext**

```typescript
"use client";
import React, { createContext, ReactNode } from "react";
import { RootStore } from "@/store/store";

export const StoreContext = createContext(RootStore);

export const StoreWrapper = ({ children }: { children: ReactNode }) => {
  return (
    <StoreContext.Provider value={RootStore}>{children}</StoreContext.Provider>
  );
};
``` 


**RootStore**

```typescript
import { UserStore } from "@/store/user-store";
import { UiStore } from "@/store/ui-store";
import { AccountStore } from "@/store/account-store";

const uiStore = new UiStore();
const userStore = new UserStore(uiStore);
const accountStore = new AccountStore(uiStore);
export const RootStore = {
  uiStore,
  userStore,
  accountStore,
};
``` 

Root layout  / page

```typescript
export default function RootLayout({ children }: { children: ReactNode }) {
  return (
    <html lang="en" className="w-full h-full">
      <body className="h-full w-full">
        <StoreWrapper>{children}</StoreWrapper>
      </body>
    </html>
  );
}

``` 


Use in component

```typescript
"use client";
import { ReactNode } from "react";
import { observer } from "mobx-react";
import { useStores } from "@/hooks/useStore";
import { Header } from "@/components/header";

const HomePage = observer(({ children }: { children: ReactNode }) => {
  const {
    uiStore: { isLoading },
  } = useStores();

  if (isLoading) {
    return <div>Loading ...</div>;
  }

  return (
    <>
      <Header />
      {children}
    </>
  );
});
export default HomePage;

``` 

Note : You need to wrap component with `observer` if you want to get real time changes else you need to render component to get updated values .

And you can create your own stores and initialize in root store and share in whole app by context api

**Reply by NileshJain78** [reply]
what about the ssr because if the component is marked as use client then it will refresh itself

**Comment by ideaviewes** [comment]
Thank you so much for your response. My project still isn't meeting expectations. Could you please take a look? I would greatly appreciate it.
https://github.com/ideaviewes/mobx-learn.git

**Reply by mawaisismail** [reply]
`import {makeAutoObservable, makeObservable} from 'mobx';`
change `makeObservable` to `makeAutoObservable`

**Reply by ideaviewes** [reply]
It works Thank you so much！！

**Reply by el-frontend** [reply]
That works for me too, thank you!!!

**Reply by sungjun-finda** [reply]
I'm trying to handle same problem, but I encouter the new issue that my stores are mostly use subclassing or super pattern so I cannot change my all model to makeAutoObservable. (also using many annotation style)

So I wonder that...
1. Why changing model to makeAutoObservable can be worked?
2. Is there another way to handle this problem...?

**Comment by wertyga** [comment]
Do you understand that your code will work only on client?

**Comment by KaneLabs** [comment]
Not only will it only work on the client, it still needs SSR.

**Comment by maksymKhaiuk1989** [comment]
I tried with latest next js with app router, with following libs, all works totaly fine, without any providers
```
     "mobx": "^6.13.7",
    "mobx-react-lite": "^4.1.0",
    "next": "latest",
```


store

```
import { makeAutoObservable } from "mobx";

class MobxStore {
  count = 0;
  constructor() {
    makeAutoObservable(this);
  }

  increment = () => {
    this.count += 1;
  };

  decrement = () => {
    if (this.count > 0) {
      this.count -= 1;
    }
  };
}

export const mobxStore = new MobxStore();
```

component
```
import { observer } from "mobx-react-lite";
import { mobxStore } from "@/lib/mobx";

const Counter = observer(() => {
  const { count, increment, decrement } = mobxStore;
  return (
    <div>
      <h1>
        Count: <span>{count}</span>
      </h1>
      <button onClick={increment}>+1</button>
      <button onClick={decrement}>-1</button>
    </div>
  );
});

```

**Reply by ai-songee** [reply]
Thank you for your approach! It is working well in a single page app which only have 1 store. 

But the store is initialized globally here and will not be disposed forever, it will become an issue when you add more pages and stores to your app. If you have multiple pages and stores, don't forget to manage the lifecycle of each store, dispose them when the view is unmounted. The context provider approach mentioned above should be a good solution.
